prov/shm/src/smr_ep.c

```diff
@@ -599,13 +729,26 @@ static void smr_generic_format(struct smr_cmd *cmd, fi_addr_t peer_id, void *con
 	if (flags & FI_TAGGED) {
 		cmd->hdr.op.op = ofi_op_tagged;
 		cmd->hdr.op.tag = tag;
+	} else if (flags & FI_WRITE) {
+		cmd->hdr.op.op = ofi_op_write;
+	} else if (flags & FI_READ) {
+		cmd->hdr.op.op = ofi_op_read_req;
 	} else {
 		cmd->hdr.op.op = ofi_op_msg;
 	}
```

Comment:
I would look at trying to set the op earlier in the flow.  I'm guessing you already know what operation to perform (based on setting the flags), and can avoid this if-else-if flow.

Reply:
Updated to included extra patch for this